components/ai_chat/core/browser/ai_chat_database.cc

```diff
@@ -0,0 +1,914 @@
+             << db_.GetErrorMessage();
+    return AddConversationResult{};
+  }
+
+  AddConversationResult result;
+
+  if (conversation->associated_content->is_content_association_possible) {
+    DVLOG(2) << "Adding associated content for conversation "
+             << conversation->uuid << " with url "
+             << conversation->associated_content->url->spec();
+    result.associated_content_id = AddOrUpdateAssociatedContent(
+        conversation->uuid, std::move(conversation->associated_content),
+        contents);
+    if (result.associated_content_id == -1) {
+      return result;
+    }
+  }
+
+  if (!AddConversationEntry(conversation->uuid, std::move(first_entry))) {
+    return AddConversationResult{};
+  }
+
+  if (!transaction.Commit()) {
+    DVLOG(0) << "Transaction commit failed with reason: "
+             << db_.GetErrorMessage();
+    return AddConversationResult{};
+  }
+
+  return result;
+}
+
+int32_t AIChatDatabase::AddOrUpdateAssociatedContent(
+    std::string_view conversation_uuid,
+    mojom::SiteInfoPtr associated_content,
+    std::optional<std::string> contents) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  // TODO(petemill): handle multiple associated content per conversation
+  CHECK(!conversation_uuid.empty());
+
+  sql::Statement statement;
+  int index;
+  if (associated_content->id.value_or(-1) != -1) {
+    DVLOG(4) << "Updating associated content for conversation "
+             << conversation_uuid << " with id "
+             << associated_content->id.value();
+    static const std::string kUpdateAssociatedContentQuery =
+        "UPDATE associated_content"
+        " SET title = ?,"
+        " url = ?,"
+        " content_type = ?,"
+        " last_contents = ?,"
+        " content_used_percentage = ?,"
+        " is_content_refined = ?"
+        " WHERE id=?";
+    statement.Assign(GetDB().GetUniqueStatement(kUpdateAssociatedContentQuery));
+    statement.BindInt(6, associated_content->id.value());
+    index = 0;
+  } else {
+    DVLOG(4) << "Inserting associated content for conversation "
+             << conversation_uuid;
+    static const std::string kInsertAssociatedContentQuery =
+        "INSERT INTO associated_content(conversation_uuid, title, url,"
+        " content_type, last_contents, content_used_percentage,"
+        " is_content_refined)"
+        " VALUES(?, ?, ?, ?, ?, ?, ?) ";
+    statement.Assign(GetDB().GetUniqueStatement(kInsertAssociatedContentQuery));
+    statement.BindString(0, conversation_uuid);
+    index = 1;
+  }
+  CHECK(statement.is_valid());
+  BindAndEncryptOptionalString(statement, index, associated_content->title);
+  index++;
+  BindAndEncryptOptionalString(statement, index,
+                               associated_content->url->spec());
+  index++;
+  statement.BindInt(index,
+                    static_cast<int32_t>(associated_content->content_type));
+  index++;
+  BindAndEncryptOptionalString(statement, index, contents);
```

Comment:
We won't cap the database, although it would be good to state how much space it's using in the "Delete browsing data" dialog.

It's going to use space only when the user makes a conversation. We could introduce a warning as the feature gets used. But a rough estimate would be that 10,000 messages would take up 5Mb and 100 conversations each associated with content with a max length of ~10,000 characters would take a total of 1Mb, unless my foggy brain is mistaken?

Reply:
Doesn't that 10,000 chars take around 10Mb(9.8Mb)? And reminds me where the max length of ~10000 restriction imposed when storing the content.

Reply:
10,000 characters will take up **10Kb**, no? We could double that estimation for encryption.

There is no cap at the moment, I'm suggesting we add one to storage and also page content fetcher.